### PR TITLE
[MH2] Implemented Carth the Lion

### PR DESCRIPTION
**Pull request by weirddan455**
Issue #7807

I'm not entirely sure how the rules are supposed to work for modifying loyalty ability costs.  The way I've implemented it, a +1 loyalty ability changes to a +2, a -1 changes to a 0, etc.

I've tested and this doesn't allow planeswalkers to ultimate early (ex. a planeswalker that has a -7 ability cannot ultimate on 6 loyalty even though it would only cost -6 loyalty with Carth in play.)  Again, not sure how this is supposed to work.  Let me know if this is not correct and I'll try to fix it.

### CODE REVIEW

**Comment by emerald000** [comment]
This adds a second cost of +1 to activating a planeswalker, similiar to how Suppression Field would add a {2} cost.

For the example about the ultimate: If you are on 6, and try to activate a -7, you have two cost to pay: -7 and +1. If you pay the +1 first, you now have 7 counters which you can remove for the -7.

**Comment by theelk801** [comment]
there's actually a [change coming](https://twitter.com/Dunkatog/status/1397661988335996928) for this

**Comment by weirddan455** [comment]
OK, so maybe just leave this open until the rule change gets released?

For what it's worth, I originally tried just adding an additional PayLoyaltyCost of 1 to the ability like you described and it resulted in the ability just not doing anything.

**Comment by theelk801** [comment]
well we know the intended functionality, we just need to modify the loyalty cost rather than adding to it

**Comment by theelk801** [comment]
the hard part is gonna be handling X costs

**Comment by weirddan455** [comment]
> we just need to modify the loyalty cost rather than adding to it

Sorry, a little confused on what you mean here.  That's what I ended up doing in this PR is modifying the existing loyalty cost.

I tested with X costs and it seems to work (again with the caveat that it won't let you set X to be greater than the number of loyalty counters currently on the planeswalker.)  With this code, it correctly removes X - 1 counters.

**Comment by theelk801** [comment]
oh ok, I hadn't read the code yet, looks good to me

**Comment by weirddan455** [comment]
OK, I just read the rest of that Twitter thread.  I'll try to fix the issue where it's not allowing a -3 to be activated with 2 loyalty counters.  I assuming this also means you should be able to declare an X of -3 with 2 loyalty counters?  Or is it different with X costs?

**Comment by theelk801** [comment]
yeah you should be able to choose X=3 with only two counters

**Comment by weirddan455** [comment]
I added functionality to check for cost modification in canPay and getMaxValue of the loyalty costs.  I've tested that this works as expected now.  Let me know if this solution looks good to you or if there's a better way to handle this.

**Comment by theelk801** [comment]
great, I'm gonna check it out in the morning

**Comment by JayDi85** [comment]
Please, try to [make unit tests](https://github.com/magefree/mage/wiki/Development-Testing-Tools) for game engine changes or new features.

File: Mage/src/main/java/mage/abilities/costs/common/PayVariableLoyaltyCost.java [hunk 1]
```diff
@@ -16,6 +17,8 @@
  */
 public class PayVariableLoyaltyCost extends VariableCostImpl  {
 
+    public int costModification = 0;
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/costs/common/PayVariableLoyaltyCost.java:20]: missing copy code

File: Mage/src/main/java/mage/abilities/costs/common/PayLoyaltyCost.java [hunk 2]
```diff
@@ -33,8 +34,19 @@ public PayLoyaltyCost(PayLoyaltyCost cost) {
 
     @Override
     public boolean canPay(Ability ability, Ability source, UUID controllerId, Game game) {
+        int loyaltyCost = amount;
+        if (ability instanceof LoyaltyAbility) {
+            LoyaltyAbility copiedAbility = ((LoyaltyAbility) ability).copy();
+            game.getContinuousEffects().costModification(copiedAbility, game);
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/costs/common/PayLoyaltyCost.java:40]: Cost modification code must be called as two methods for compatibility (cause cost modification possible from the source/itself and from another effects):
```
sourceObject.adjustCosts(copyAbility, game);
game.getContinuousEffects().costModification(copiedAbility, game);
```

File: Mage/src/main/java/mage/abilities/costs/common/PayLoyaltyCost.java [hunk 3]
```diff
@@ -33,8 +34,19 @@ public PayLoyaltyCost(PayLoyaltyCost cost) {
...
+            loyaltyCost = 0;
+            for (Cost cost : copiedAbility.getCosts()) {
+                if (cost instanceof PayLoyaltyCost) {
+                    loyaltyCost += ((PayLoyaltyCost) cost).amount;
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/costs/common/PayLoyaltyCost.java:44]: All loyalty cost modification must be moved to standard method with normal cost modification:
* `public void costModification(Ability abilityToModify, Game game)`

Or maybe there were a reason for your workaround (for special loyalty modification code)?

File: Mage/src/main/java/mage/abilities/costs/common/PayLoyaltyCost.java [hunk 4]
```diff
@@ -16,7 +17,7 @@
  */
 public class PayLoyaltyCost extends CostImpl {
 
-    private final int amount;
+    public int amount;
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/costs/common/PayLoyaltyCost.java:20]: Make property, not public variable. Example: setAmount, getAmount.

File: Mage/src/main/java/mage/abilities/costs/common/PayLoyaltyCost.java (hunk 3 above)
Comment (weirddan455) [reply Mage/src/main/java/mage/abilities/costs/common/PayLoyaltyCost.java:44]: Not sure what you mean here.  The issue is that the costModification method wasn't getting called before canPay does so I called it in line 40.

All this code you quoted does is get the new cost (after modification) and uses it to check if the planeswalker has enough loyalty counters.

**Comment by Grath** [comment]
Any progress on this? Carth is one of the cards I was looking forward to the most with MH2 being released.

**Pull request merged**